Update CONTRIBUTING.md

ThomasLandauer opened the pull request:
I shortened the "preamble" a bit :-)

The `====` line is just my marker where I stopped ;-)

Question: How is the syncing between https://github.com/Codeception/module-symfony and https://github.com/Codeception/symfony-module-tests happening? Automatically, I guess? On each *merge* or on each *release*?

And am I right in my assumption that the benefit of https://github.com/Codeception/symfony-module-tests is saving everybody from having to set up their own Symfony project? Or anything else?

TavoNiievez commented:
> On each merge or on each release?

@ThomasLandauer It is done by releases.

Each release includes the tests of the methods that were added.

> Question: How is the syncing between Codeception/module-symfony and Codeception/symfony-module-tests happening? Automatically, I guess?

It is not automatic, in the sense that someone must write the tests for the functions that are added, or update the Symfony version when there is a new release *(eg 5.3 or 5.4)*

> the benefit of Codeception/symfony-module-tests is saving everybody from having to set up their own Symfony project? Or anything else?

Well, it has 4 main utilities:

- The first is to verify that the changes that are added with a PR will not break or alter the result of the current tests.
- The second is to verify that the same test code works in several versions of Symfony, that is why Codeception/symfony-module-tests has 3 branches, one for each stable version of Symfony, with 'main' being the last stable version available, which at this moment is `5.2`
- Another advantage is to avoid that anyone has to create a project from scratch to reproduce a bug or to add a new function. The application code is minimal, therefore it is more reliable than writing it in a larger pre-existing application, because some bundle or added configuration can add 'noise' or unexpected behavior in the tests.
- And the last advantage is probably 'practical documentation'. Well, it shows in code how to use the functions in a real application

TavoNiievez commented on CONTRIBUTING.md:
```diff
@@ -1,11 +1,10 @@
+First of all: Contributions are very welcome!
 
-If instead you plan to edit or add some new function to this module, hey, you rock!
+For *larger* changes, we recommend to use the test project for this module: [Codeception/symfony-module-test](https://github.com/Codeception/symfony-module-tests). It is a minimal (but complete) Symfony project, ready to run tests. The code inside `vendor/codeception/module-symfony` is the same as...
```
OK, but how would i know how 'long' my changes are to choose to use the test project vs the Github editor?

You should test your change in the test project if you are going to modify the internal code of a function or you will add a totally new one, but if you will only add comments, or change the name of some variable it may not be so necessary.

ThomasLandauer replied:
See today's Commit :-)

ThomasLandauer commented:
Sorry, just noticed that I had a completely wrong assumption about those projects' interplay!
1. The tests at https://github.com/Codeception/symfony-module-tests/tree/main/tests/Functional are not included in the module itself? So there is no way to run these tests in your "regular" project? So if you do a PR that affects the tests, you **have to** download https://github.com/Codeception/symfony-module-tests ?
1a. When are the tests actually run? You probably hooked them into the CI release process on GitHub somehow?
2. The code of https://github.com/Codeception/module-symfony/blob/master/src/Codeception/Module/Symfony.php is not included in https://github.com/Codeception/symfony-module-tests ? It's just installed by composer, same as is any "real" project?

TavoNiievez commented:
> The tests at https://github.com/Codeception/symfony-module-tests/tree/main/tests/Functional are not included in the module itself? 

Right.

So there is no way to run these tests in your "regular" project?

Yes you can, by editing the vendor files of your project, but as i said, that adds many variables and *'custom settings'* that can affect the test result in other environments.

So if you do a PR that affects the tests, you have to download Codeception/symfony-module-tests ?

You don't **have to**, but you **should**.

> When are the tests actually run? You probably hooked them into the CI release process on GitHub somehow?

they are executed with each PR and with each merge. See https://github.com/Codeception/module-symfony/blob/master/.github/workflows/main.yml

> The code of https://github.com/Codeception/module-symfony/blob/master/src/Codeception/Module/Symfony.php is not included in Codeception/symfony-module-tests ?

No.

It's just installed by composer, same as is any "real" project?

Yes.

ThomasLandauer commented:
> Yes you can, by editing the vendor files of your project

But how can I run the tests? - They just aren't there... I'd have to manually copy the contents of https://github.com/Codeception/symfony-module-tests/blob/main/tests/Functional/SymfonyModuleCest.php into some file in my project?

TavoNiievez commented:
Ahhh if you mean running ***those*** tests then no. I only mean that it is possible to execute your own new functions or changes in your existing project.

ThomasLandauer commented:
OK :-) See https://github.com/TavoNiievez/module-symfony/pull/1/commits/0b646a06fc05d66fd1282d7964b679ff27e46958

TavoNiievez reviewed (approved):
Great! All good so far!

TavoNiievez commented:
@ThomasLandauer What do you need to know to continue?

TavoNiievez commented on CONTRIBUTING.md:
```diff
@@ -1,16 +1,35 @@
+2. Edit the modules source code in `vendor/codeception/module-symfony/src/Codeception/Module/Symfony.php` of this project.
+
+3. Create a test with the same name as your new function in `tests/Functional/SymfonyModuleCest.php`, following alphabetical order.
+
```
Here would come the explanation of `vendor/bin/codecept clean` and `vendor/bin/codecept build`

TavoNiievez commented on CONTRIBUTING.md:
```diff
@@ -1,16 +1,35 @@
+
+4. Run the tests with `vendor/bin/codecept run Functional`.
+
+5. ???
```
anything else that needs clarification?

ThomasLandauer commented:
I posted some questions at https://github.com/TavoNiievez/module-symfony/pull/1/commits/0b646a06fc05d66fd1282d7964b679ff27e46958 - it looks like you didn't notice them ;-)

Questions:

1. Do you agree in principle with the case separation (test or no test required)?
2. What did you mean with "and verify that it works correctly with various test data."? Which test data?
3. The `vendor/bin/codecept clean` part is a hint to make life easier in step 2?
4. In the end (i.e. at step 5) you have to send a PR to *two*  repos, right?
5. `git clone https://github.com/YourUserName/module-symfony.git` only works *after* you forked the project? Forking can only be done on GitHub's website? (i.e. not through the command line)?
6. Isn't `git add -A` missing before `git commit`?

TavoNiievez commented:
> I posted some questions at 0b646a0 - it looks like you didn't notice them ;-)

Sorry, it's a very unusual place to ask, but it's okay :)
<hr/>

> Do you agree in principle with the case separation (test or no test required)?

Yes, although i would have preferred the question to guide me and not the answer. For example:
 - Do you add or modify functionality?
 **Yes.** Then you will need to add some tests...
 - Is your change about documentation, or just a really minor code change?
 **Yes.** Then...
<hr/>

> What did you mean with "and verify that it works correctly with various test data."? Which test data?

Check this code:

https://github.com/Codeception/symfony-module-tests/blob/b57b9fc2c406555b58d480d688e4ae5849edf1b4/tests/Functional/SymfonyModuleCest.php#L92-L112

That method, for example, works with several types of input data as arguments, and tests should test each of them.
<hr/>

> The vendor/bin/codecept clean part is a hint to make life easier in step 2?

I recommend cleaning and rebuilding the Tester Actions when you are editing or adding methods. In theory just `codecept build` would work, but just to be safe i run a `codecept clean` first.
<hr/>

> In the end (i.e. at step 5) you have to send a PR to two repos, right?

That's right.
<hr/>

> `git clone https://github.com/YourUserName/module-symfony.git` only works after you forked the project? 

Yes, you must clone your fork, because if you clone the Codeception repository you will not be able to make commits since you do not have permissions to do so.
<hr/>

> Forking can only be done on GitHub's website? (i.e. not through the command line)?

Forking is a GitHub feature and does not come from Git. In theory there is `GitHub CLI` that allows you to do it by command but not everyone has it and we should not request that it be installed here since it is only to click on the fork button on the web.
<hr/>

> Isn't git add -A missing before git commit?

No, the `Codeception/Module/Symfony.php` file is already being tracked by Git. You would only need -A if you were to create a totally new file. 

Maybe what i did omit was to add a message for that commit, for example: `git commit -m "Adding dontSeeInSession"`

TavoNiievez commented on CONTRIBUTING.md:
```diff
@@ -13,10 +13,19 @@ Does your change require a test?
 2. Edit the modules source code in `vendor/codeception/module-symfony/src/Codeception/Module/Symfony.php` of this project.
 
 3. Create a test with the same name as your new function in `tests/Functional/SymfonyModuleCest.php`, following alphabetical order.
+  Hint: Run this to rebuild Codeception's "Actor" classes (see [Console Commands](https://codeception.com/docs/reference/Commands#Build)) to get auto-completion in your IDE:
```
Cool!

ThomasLandauer replied:
:-)

ThomasLandauer commented:
> Yes, although i would have preferred the question to guide me and not the answer.

I wanted to (re)use the important part as heading (in case somebody doesn't read all the fine print). But if you have a better idea => sure :-)

> "various test data"

I'm omitting this, since it's a special case. I think we can/must rely on people knowing by themselves what needs to be tested.

Questions:
1. In step 5, I'd say it's better to send the *test* project PR first (to *finish* this, before forking the actual module). What are the git commands?
2. How will you (=maintainers) know which test PR belongs to which module PR? Idea: Mention the (just created) test PR in the following module PR.
3. At `git branch <name_of_your_new_branch>` does the name matter or is something like "patch-1" sufficient?
4. The very last step ("Finally go back to GitHub and create a Pull Request from...") can't be done on CLI?
5. Starting from step 6 is *exactly* the same as the case "No, My Change Does Not Require a Test", right?
6. Starting from step 6: What about just doing it on GitHub's website, and pasting (i.e. overwriting) the **entire** file `Symfony.php`?

TavoNiievez commented:
> In step 5, I'd say it's better to send the test project PR first (to finish this, before forking the actual module). What are the git commands for this?

Hmm... I don't think so, you should first send the change to `Codeception/module-symfony` and then, when it is merged in `master` send the PR to the Test Project, the reason for this is that some of the maintainers can request changes in your method or directly reject it. Similarly, only if your PR is merged to `master` in `module-symfony` will you be able to use it in the Test Project, otherwise the CI job will fail.

> How will you (=maintainers) know which test PR belongs to which module PR? Idea: Mention the (just created) test PR in the following module PR's commit message. Is there a key word (like "closes") for this?

I think the solution for this is to create PR templates, so that information like this is filled in. Like:

Q | A
-- | --
Bug fix? | no
New feature? | yes
Documentation? | no
Tests PR | ---

> At git branch <name_of_your_new_branch> does the name matter or is something like "patch-1" sufficient?

Using the default names is bad practice, even if you will only make one change you should try to add a descriptive name, however it is not mandatory.

> The very last step ("Finally go back to GitHub and create a Pull Request from...") can't be done on CLI?

PRs are also a feature of GitHub, you can only do it with the `GitHub CLI`. So... no.

> Starting from step 6 is exactly the same as the case "No, My Change Does Not Require a Test", right?

Yes.

> Starting from step 6: What about just doing it on GitHub's website, and pasting (i.e. overwriting) the entire file Symfony.php?

Even if you are using the GitHub editor you should not copy and paste the entire content of the file, there may be some problems with the spacing that are detected as change and corrupt the git history.

You should only overwrite the piece of code you are going to edit.

TavoNiievez commented:
> What are the git commands for this?

Exactly the same as with the module: Fork the project, clone it locally, make the change, do a git commit, do a git push, and create the PR on the Web.

ThomasLandauer commented:
> you should first send the change to Codeception/module-symfony

OK, that makes sense. But then we need to adapt the overall workflow:
1. Install test project
2. Do the module code change and PR it
3. **Wait for feedback**
4. Write the test

PR template: I think this would over-complicate it. With the new workflow (see above), don't you think that just telling people to mention the corresponding PR in step 4 would be enough?

TavoNiievez commented:
> don't you think that just telling people to mention the corresponding PR in step 4 would be enough?

Yes, you're right.

What causes me confusion in your steps is that for me to write a function and send a PR, I had to have previously tested that this function works, right?

So probably:
```diff
- 4. Write the test
+ 4. Send the test
```

ThomasLandauer commented:
* I changed `git branch new_feature` to `git checkout -b new_feature`
* I added `git add -A` cause without it I was getting:
   > no changes added to commit (use "git add" and/or "git commit -a")

TavoNiievez commented on CONTRIBUTING.md:
````diff
@@ -1,49 +1,76 @@
+
+5. Commit:
+  ```bash
+  git add -A
+  git commit -m "Briefly explain what your change is about"
````
```suggestion
  git commit --all --message="Briefly explain what your change is about"
```

TavoNiievez replied:
if the` git add` is mandatory then just replace `-a` with `--all`

ThomasLandauer commented:
I'm done :-) Please double-check everything, especially the git commands...

TavoNiievez commented on CONTRIBUTING.md:
````diff
@@ -1,49 +1,70 @@
 
+9. Commit:
+  ```bash
+  git add -A
+  git commit -m "Add a link to the module's Pull Request you created above"
````
```suggestion
  git add --all
  git commit --message="Add a link to the module's Pull Request you created above"
```

ThomasLandauer commented:
> What causes me confusion in your steps is that for me to write a function and send a PR, I had to have previously tested that this function works, right?

I'd rather wait for some feedback on the idea in general, before writing a polished test ;-) But ultimately it's up to the user, in which order they want to do it - that's why I added "If you want, you can already write the tests"

ThomasLandauer reviewed (commented):
:+1:

TavoNiievez commented on CONTRIBUTING.md:
```diff
@@ -1,49 +1,70 @@
+  git clone https://github.com/YourUserName/symfony-module-tests.git
+  cd symfony-module-tests
+  git checkout -b new_feature
+  composer update
```
we shouldn't include this step, it depends on the php version the user is running and is already explained in the test project readme.md
```suggestion
```

TavoNiievez replied:
I'm going to merge this without the `composer update`
